src/org/jitsi/impl/neomedia/rtp/translator/RTCPFeedbackMessageSender.java

```diff
@@ -102,87 +113,287 @@ private long getSenderSSRC()
 
     /**
      * Sends a Full Intra Request (FIR) Command to a media sender/source with a
-     * specific synchronization source identifier (SSRC).
-     * <p>
-     * <b>Warning</b>: Due to (current) limitations in
-     * <tt>RTPTranslatorImpl</tt> and/or <tt>StreamRTPManager</tt>, a FIR
-     * command with the specified <tt>mediaSenderSSRC</tt> is sent to all
-     * <tt>MediaStream</tt>s.
-     * </p>
+     * specific synchronization source identifier (SSRC) through a specific
+     * <tt>MediaStream</tt>.
      *
      * @param mediaSenderSSRC the SSRC of the media sender/source
      * @return <tt>true</tt> if a FIR command was sent; otherwise,
      * <tt>false</tt>
      */
     public boolean sendFIR(int mediaSenderSSRC)
     {
-        boolean sentFIR = false;
-        /*
-         * XXX Currently, the method effectively broadcasts a FIR i.e. sends it
-         * to all streams connected to rtpTranslator because MediaStream's
-         * method getRemoteSourceIds returns an empty list (possibly due to
-         * RED).
-         */
-        for (StreamRTPManager streamRTPManager
-                : rtpTranslator.getStreamRTPManagers())
-        {
-            MediaStream stream = streamRTPManager.getMediaStream();
+        // It's OK for this method to be a little "slow" (so that the
+        // {@code RTCPFeedbackMessageSender#maybeStopRequesting} is lock-less).
+        KeyframeRequester oldRequester = kfRequesters.putIfAbsent(
+            mediaSenderSSRC, new KeyframeRequester(mediaSenderSSRC));
```

Comment:
I wanted to use a ConcurrentHashMap instead of a synchronized block (which would also have to be present in the `maybeStopRequesting` that runs for every incoming RTP packet). My thinking was that sending FIRs is much less often than RTP packets, so it's okay for the sendFIR method to be slow. I don't mind changing this, if you prefer it the other way.